utils.go

```diff
@@ -1,27 +1,108 @@
 package ipfs
 
 import (
+	"errors"
+	"fmt"
+
+	ipfs "github.com/ipfs/go-ipfs-api"
+
+	"github.com/beyondstorage/go-endpoint"
+	"github.com/beyondstorage/go-storage/v4/services"
 	"github.com/beyondstorage/go-storage/v4/types"
 )
 
 // Storage is the example client.
 type Storage struct {
+	ipfs *ipfs.Shell
+
 	defaultPairs DefaultStoragePairs
 	features     StorageFeatures
 
+	name    string
+	workDir string
+
 	types.UnimplementedStorager
+	types.UnimplementedCopier
+	types.UnimplementedMover
 }
 
 // String implements Storager.String
 func (s *Storage) String() string {
-	panic("implement me")
+	return fmt.Sprintf(
+		"Storager IPFS {Name: %s, WorkDir: %s}",
+		s.name, s.workDir,
+	)
 }
 
 // NewStorager will create Storager only.
 func NewStorager(pairs ...types.Pair) (types.Storager, error) {
-	panic("implement me")
+	opt, err := parsePairStorageNew(pairs)
+	if err != nil {
+		return nil, err
+	}
+
+	st := &Storage{
+		name:    opt.Name,
+		workDir: "/",
+	}
+	if opt.HasWorkDir {
+		st.workDir = opt.WorkDir
+	}
+
+	ep, err := endpoint.Parse(opt.Endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	e := "http://localhost:5001"
+	switch ep.Protocol() {
+	case endpoint.ProtocolHTTP:
+		e, _, _ = ep.HTTP()
+	case endpoint.ProtocolHTTPS:
+		e, _, _ = ep.HTTPS()
+	}
+
+	sh := ipfs.NewShell(e)
+	if !sh.IsUp() {
+		return nil, errors.New("ipfs not online")
+	}
+	st.ipfs = sh
+
+	return st, nil
+}
+
+func formatError(err error) error {
+	if _, ok := err.(services.InternalError); ok {
+		return err
+	}
+
+	e, ok := err.(*ipfs.Error)
+	if !ok {
+		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
+	}
+
+	switch e.Message {
+	case "file does not exist":
+		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, err)
+	}
+
+	return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
 }
 
 func (s *Storage) formatError(op string, err error, path ...string) error {
-	panic("implement me")
+	if err == nil {
+		return nil
+	}
+
+	return services.StorageError{
+		Op:       op,
+		Err:      formatError(err),
+		Storager: s,
+		Path:     path,
+	}
+}
+
+// getAbsPath will calculate object storage's abs path
+func (s *Storage) getAbsPath(path string) string {
+	return s.workDir + path
```

Comment:
What if `path` is already an abs path? Do we need to treat it specially?

Reply:
This method is copied from an existing service.

https://github.com/beyondstorage/go-service-qingstor/blob/c0f9ce3358420980c0f0be15ef393b387fdb8503/utils.go#L365

Reply:
But for example fs treat it differently: https://github.com/beyondstorage/go-service-fs/blob/master/utils.go#L163-L174

@Xuanwo What's your idea?

Reply:
Pair `workDir` is for all operation, I think all paths passed in by the user should be subpaths of that path. 

If I'm wrong and `workDir` should be a suggestion rather than a restriction, then I can make a special treatment for absolute and relative paths separately.

By the way, if the user passes in `abc/... /... /... /dir/file`, will the underlying service receive this path as is, and do I need to restrict the paths that can be accessed if `workDir` is specified?

Reply:
> I think all paths passed in by the user should be subpaths of that path.

It's true, except for user input paths like `/abc/def`.

So all our services should support two kinds of paths:

- Relative Path: all of them are the subpath of `WorkDir`.
- Absolute Path: mostly start with `/` just like in Linux.

And `WorkDir` is always an absolute path.

We add this rule for:

- symlink support
- user friendly

Reply:
I agree that `workDir` (or the `path` parameter in operations) should have consistent behaviour, but it seems that this is not clearly specified?

Ping @beyondstorage/go-storage-leader @beyondstorage/go-storage-committer @beyondstorage/go-storage-reviewer  @beyondstorage/go-storage-contributor to take a look

Reply:
> I agree that `workDir` (or the `path` parameter in operations) should have consistent behaviour, but it seems that this is not clearly specified?

Yes, the behavior is not clearly specified.

I start an issue https://github.com/beyondstorage/specs/issues/132 and will comp up a GSP to standardize the behavior.